engine/access/rest/converter.go

```diff
@@ -3,21 +3,186 @@ package rest
 	"github.com/onflow/flow-go/model/flow"
 )
 
-// Converter provides the conversion function to/from the Swagger object to Flow objects
+// Converter provides functionality to convert from request models generated using
+// open api spec and flow models.
 
-func toBlock(flowBlock *flow.Block) *generated.Block {
+// Flow section - converting request data to flow models with validation.
+
+func toID(id string) (flow.Identifier, error) {
+	valid, _ := regexp.MatchString(`^[0-9a-fA-F]{64}$`, id)
+	if !valid {
+		return flow.Identifier{}, fmt.Errorf("invalid ID")
+	}
+
+	return flow.HexStringToIdentifier(id)
+}
+
+func toAddress(address string) (flow.Address, error) {
+	valid, _ := regexp.MatchString(`^[0-9a-fA-F]{16}$`, address)
+	if !valid {
+		return flow.Address{}, fmt.Errorf("invalid address")
+	}
+
+	return flow.HexToAddress(address), nil
+}
+
+func toProposalKey(key *generated.ProposalKey) (flow.ProposalKey, error) {
+	address, err := toAddress(key.Address)
+	if err != nil {
+		return flow.ProposalKey{}, err
+	}
+
+	return flow.ProposalKey{
+		Address:        address,
+		KeyIndex:       uint64(key.KeyIndex),
+		SequenceNumber: uint64(key.SequenceNumber),
+	}, nil
+}
+
+func toSignature(signature *generated.TransactionSignature) (flow.TransactionSignature, error) {
+	address, err := toAddress(signature.Address)
+	if err != nil {
+		return flow.TransactionSignature{}, err
+	}
+
+	return flow.TransactionSignature{
+		Address:     address,
+		SignerIndex: int(signature.SignerIndex),
+		KeyIndex:    uint64(signature.KeyIndex),
+		Signature:   []byte(signature.Signature),
+	}, nil
+}
+
+func toSignatures(sigs []generated.TransactionSignature) ([]flow.TransactionSignature, error) {
+	signatures := make([]flow.TransactionSignature, len(sigs))
+	for _, sig := range sigs {
+		signature, err := toSignature(&sig)
+		if err != nil {
+			return nil, err
+		}
+
+		signatures = append(signatures, signature)
+	}
+
+	return signatures, nil
+}
+
+func toTransaction(tx *generated.TransactionsBody) (flow.TransactionBody, error) {
+	args := make([][]byte, len(tx.Arguments))
+	for _, arg := range tx.Arguments {
+		// todo validate
+		args = append(args, []byte(arg))
+	}
+
+	proposal, err := toProposalKey(tx.ProposalKey)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	payer, err := toAddress(tx.Payer)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	auths := make([]flow.Address, len(tx.Authorizers))
+	for _, auth := range tx.Authorizers {
+		a, err := toAddress(auth)
+		if err != nil {
+			return flow.TransactionBody{}, err
+		}
+
+		auths = append(auths, a)
+	}
+
+	payloadSigs, err := toSignatures(tx.PayloadSignatures)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	envelopeSigs, err := toSignatures(tx.EnvelopeSignatures)
+	if err != nil {
+		return flow.TransactionBody{}, err
+	}
+
+	return flow.TransactionBody{
+		ReferenceBlockID:   flow.Identifier{},
+		Script:             []byte(tx.Script),
```

Comment:
Should we? Is there actually a limit for this in the protocol?

Reply:
Also, this is not correct since Script is base64 encoded in the spec

Reply:
we do if we want to avoid someone uploading an extremely large script. ANs have finite amounts of memory so we need to protect them from easy DoS attacks